federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -224,13 +240,121 @@ func New(federationClient fedclientset.Interface, dns dnsprovider.Interface,
+		},
+	}
+	fedInformerFactory := func(cluster *v1beta1.Cluster, targetClient kubeclientset.Interface) (cache.Store, cache.ControllerInterface) {
+		return cache.NewInformer(
+			&cache.ListWatch{
+				ListFunc: func(options v1.ListOptions) (pkgruntime.Object, error) {
+					return targetClient.Core().Services(v1.NamespaceAll).List(options)
+				},
+				WatchFunc: func(options v1.ListOptions) (watch.Interface, error) {
+					return targetClient.Core().Services(v1.NamespaceAll).Watch(options)
+				},
+			},
+			&v1.Service{},
+			controller.NoResyncPeriodFunc(),
+			// Trigger reconciliation whenever something in federated cluster is changed. In most cases it
+			// would be just confirmation that some service operation succeeded.
+			util.NewTriggerOnAllChanges(
+				func(obj pkgruntime.Object) {
+					// TODO: Use this to enque services.
+				},
+			))
+	}
+
+	s.federatedInformer = fedutil.NewFederatedInformer(federationClient, fedInformerFactory, &clusterLifecycle)
+
+	federatedUpdater := fedutil.NewFederatedUpdater(s.federatedInformer,
+		func(client kubeclientset.Interface, obj pkgruntime.Object) error {
+			svc := obj.(*v1.Service)
+			_, err := client.Core().Services(svc.Namespace).Create(svc)
+			return err
+		},
+		func(client kubeclientset.Interface, obj pkgruntime.Object) error {
+			svc := obj.(*v1.Service)
+			_, err := client.Core().Services(svc.Namespace).Update(svc)
+			return err
+		},
+		func(client kubeclientset.Interface, obj pkgruntime.Object) error {
+			svc := obj.(*v1.Service)
+			err := client.Core().Services(svc.Namespace).Delete(svc.Name, &v1.DeleteOptions{})
+			return err
+		})
+
+	s.deletionHelper = deletionhelper.NewDeletionHelper(
+		s.hasFinalizerFunc,
+		s.removeFinalizerFunc,
+		s.addFinalizerFunc,
+		// objNameFunc
+		func(obj pkgruntime.Object) string {
+			service := obj.(*v1.Service)
+			return service.Name
+		},
+		updateTimeout,
+		s.eventRecorder,
+		s.federatedInformer,
+		federatedUpdater,
+	)
+
 	s.endpointWorkerMap = make(map[string]bool)
 	s.serviceWorkerMap = make(map[string]bool)
 	s.endpointWorkerDoneChan = make(chan string, maxNoOfClusters)
 	s.serviceWorkerDoneChan = make(chan string, maxNoOfClusters)
 	return s
 }
 
+// Returns true if the given object has the given finalizer in its ObjectMeta.
+func (s *ServiceController) hasFinalizerFunc(obj pkgruntime.Object, finalizer string) bool {
+	service := obj.(*v1.Service)
+	for i := range service.ObjectMeta.Finalizers {
+		if string(service.ObjectMeta.Finalizers[i]) == finalizer {
+			return true
+		}
+	}
+	return false
+}
+
+// Removes the finalizer from the given objects ObjectMeta.
+// Assumes that the given object is a service.
+func (s *ServiceController) removeFinalizerFunc(obj pkgruntime.Object, finalizer string) (pkgruntime.Object, error) {
+	service := obj.(*v1.Service)
+	newFinalizers := []string{}
+	hasFinalizer := false
+	for i := range service.ObjectMeta.Finalizers {
```

Comment:
I didn't put enough focus into this code as this was nikhil's commit and was already LGTM'ed and i just verified the functionality. I would update the code as needed. This would also need to be taken care in other controllers also i believe.